[WIP] Issue 309 update admin user role

kate-kyzyma opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento-functional-tests-migration#309: Convert UpdateAdminUserRoleEntityTest to MFTF 

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

atwixfirster commented on app/code/Magento/Backend/Test/Mftf/ActionGroup/AssertUserRoleRestrictedAccessActionGroup.xml:
```diff
@@ -0,0 +1,14 @@
+    <actionGroup name="AssertUserRoleRestrictedAccessActionGroup">
+        <see selector="{{AdminHeaderSection.pageHeading}}" userInput="Sorry, you need permissions to view this content." stepKey="seeErrorMessage"/>
+    </actionGroup>
+</actionGroups>
```
@kate-kyzyma , could you please add an empty line after `</actionGroups>`.

Many thanks!

kate-kyzyma replied:
@atwixfirster fixed.

Thanks.

rogyar commented on app/code/Magento/Backend/Test/Mftf/ActionGroup/AssertUserRoleRestrictedAccessActionGroup.xml:
```diff
@@ -0,0 +1,14 @@
+
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="AssertUserRoleRestrictedAccessActionGroup">
```
Please, use the area name after `Assert` prefix
```suggestion
    <actionGroup name="AssertAdminUserRoleRestrictedAccessActionGroup">
```
Please, check other places as well

kate-kyzyma replied:
Fixed.

rogyar commented:
Hi @kate-kyzyma. Thank you for your contribution. I would like to explain the way how we can improve the approach used in the test. Basically, some tests require a set pre-conditions in order to execute operations. For example, if you test some particular functionality for Customer, you need to have this customer in the database - that's the main precondition.
Ideally, the preconditions should be met before the test body execution. If we execute preconditions inside a test, we delegate a lot of additional heavy and slow operations that don't test anything but create/prepare/etc. 
In this particular case, we need to have an admin user and all roles created before we run the test. 
We already have meta for [admin user](https://github.com/magento/magento-functional-tests-migration/blob/5dcb21b1d7ac9179e2d426d7b2f3750159b0d1b5/app/code/Magento/User/Test/Mftf/Metadata/user-meta.xml#L8) and for [admin role](https://github.com/magento/magento-functional-tests-migration/blob/1407b93a6d450c22c96b5829dcb10809d78035e8/app/code/Magento/User/Test/Mftf/Metadata/user_role-meta.xml#L8)

Here is an [example](https://github.com/magento/magento-functional-tests-migration/blob/5dcb21b1d7ac9179e2d426d7b2f3750159b0d1b5/app/code/Magento/Security/Test/Mftf/Test/AdminUserLockWhenEditingUserTest.xml#L28) of how to create an entity for test. The entity is declared [here](https://github.com/magento/magento-functional-tests-migration/blob/5dcb21b1d7ac9179e2d426d7b2f3750159b0d1b5/app/code/Magento/User/Test/Mftf/Data/UserData.xml#L19) using the corresponding meta as `type`.

So, the main idea of the improvement is to create the user and role in the `<before>` section of the test using the `createData` as described above.

rogyar commented on app/code/Magento/User/Test/Mftf/Data/UserRoleData.xml:
```diff
@@ -13,15 +13,29 @@
         <data key="resourceAccess">All</data>
         <data key="resources">[]</data>
     </entity>
+    <entity name="adminRole" type="role">
```
Could we try to reuse the existing entities instead of introducing new? For the current test, we need to check that roles assignment functionality is working. So, basically, we need a couple of different roles for the test. And we already have this "couple of different roles" such as `roleSales` (with restricted resources), `roleAdministrator` and so on. 
So we already have roles with different names and resource in the system. It's a good idea to reuse them. Thank you.

kate-kyzyma replied:
Fixed.

kate-kyzyma commented:
@rogyar thank you for the detailed explanation of using createData. I have edited the test. Could you, please, check it one more time?

Thanks.

rogyar commented on app/code/Magento/User/Test/Mftf/Data/UserRoleData.xml:
```diff
@@ -78,4 +78,227 @@
         </array>
     </entity>
+
+    <entity name="roleAdmin" type="user_role">
```
Could we use an existing role from `app/code/Magento/User/Test/Mftf/Data/UserRoleData.xml` instead of creating one? I'm just asking because we have a set of predefined roles and it looks like we can use them instead of creating a new one. But I might be wrong.

kate-kyzyma replied:
Hello, @rogyar 
Actually, we have only one predefined role (adminWithoutBulkActionRole) with the needed type "user_role", in that file, (for creating roles via CreateData we should use this data type). This role has restricted permissions so I am not able to navigate to required pages (for example, admin/user_role/). Tha's why I have identified a new role with full access.

Thanks.

rogyar commented on dev/tests/acceptance/.credentials:
```diff
@@ -0,0 +1 @@
+MAGENTO_ADMIN_PASSWORD=Yy111111
```
I believe we don't need this file here

kate-kyzyma replied:
The file was removed.

rogyar commented on app/code/Magento/User/Test/Mftf/Test/AdminUpdateUserRoleEntityTest.xml:
```diff
@@ -0,0 +1,130 @@
+        </before>
+
+        <after>
+            <actionGroup ref="logout" stepKey="logOut"/>
```
Please note, if you create an entity in the `before` section, you should remove it in the `after` section, not in the test's body. Please, take a look at the following example https://github.com/magento/magento-functional-tests-migration/blob/e6bbd096492264056e21bc71d4ef93cc56f8e6f5/app/code/Magento/Customer/Test/Mftf/Test/StorefrontResetCustomerPasswordFailedTest.xml#L27

kate-kyzyma replied:
I have edited the test. Now, user and role entities are removing in `after` section.

I used this test as an example for deleting the role entity in `after` section:
https://github.com/magento/magento-functional-tests-migration/blob/484cc980c79538155833f6b077b11864280069e6/app/code/Magento/User/Test/Mftf/Test/AdminBulkOperationsLogIsNotAccessibleForAdminUserWithLimitedAccessTest.xml#L35

rogyar commented:
Hi @kate-kyzyma. Thank you for the adjustments. Please, check my recent comments. 
Thank you

soumyau commented:
Closing and reopening the PR to refresh Adobe CLA sign.

kate-kyzyma commented:
Hi, @rogyar  and @soumyau 

I have merged 2.4-develop branch into the working branch and resolved merge conflict. I also updated my test a little bit: split the main test into 3 separate tests, removed my own "roleAdmin" role and reused existing roles from app/code/Magento/User/Test/Mftf/Data/UserRoleData.xml. 

Unfortunately, my tests failed for EE and B2B checks. From the reports I see that it caused by AdminDeleteRoleByRoleNameActionGroup. It is one of the existing groups which I reused in my test and I have no clue why it doesn't work.
Also, some other tests were failed app/code/Magento/Sales/Test/Mftf/Test/AdminCreateCreditMemoBankTransferPaymentTest.xml and app/code/Magento/Sales/Test/Mftf/Test/AdminMassOrdersCancelProcessingAndClosedTest.xml. I tried to re-run Functional tests for B2B and EE but it didn't help.

I don't have any issue locally and it works well for CE check. Could you help me correct my tests for EE and B2B?

Thanks.
